### ISSUE TITLE
git-style subcommands mess up argument order when using boolean flags

### ISSUE THREAD

**Issue by mal** [tj/commander.js#508]
**index.js**

``` js
var command = require('commander')

var cmd = command
  .command('hello', 'description')
  .parse(process.argv)
```

**index-hello.js**

``` js
var command = require('commander')

var cmd = command
  .option('--dry-run', 'prints actions, but takes none')
  .parse(process.argv)

console.log('opts', cmd.opts())
console.log('args', cmd.args)
```

**output**

```
$ node index.js hello 1 2 --dry-run 3 4 5 6
opts { dryRun: true }
args [ '1', '2', '4', '5', '6', '3' ]
```

Note how `3` has been shifted to the end. This really messes with you when working with subcommands that expect a specific argument ordering.

_Found and tested against `commander@2.9.0`_

**Comment by jmatsushita** [comment]
Ran into the same problem just now.

**Comment by nibblesnbits** [comment]
I'm having the same issue.  It may have to do with the fact that even `process.argv` does not contain the flags passed in.

I have a simple cli with this definition in `app`:
```js
#!/usr/bin/env node

var program = require('commander');
var package = require('./package.json');

program
  .version(package.version)
  .option('-f, --force', 'skip confirmation')
  .command('flush <queue>', 'flush the queue of all jobs');
  
program.parse(require('process').argv);
```
And this in `app-flush.js`:
```js
console.log(JSON.stringify(process.argv));
console.log(JSON.stringify(program));
```

But the output of `./app flush -f` is:

```
["C:\\Program Files\\nodejs\\node.exe","C:\\dev\\app\\app-flush"]
{"commands":[],"options":[],"_execs":{},"_allowUnknownOption":false,"_args":[],"_name":""}
```
It seems that nothing is catching that flag.

**Comment by nukelytide** [comment]
I can confirm this issue as well.

**Comment by stnever** [comment]
I can confirm this issue on commander@2.18.0 as well.

**Comment by shadowspawn** [comment]
This was reported again in #962 which has some good reproduction examples and discussion.

**Comment by shadowspawn** [comment]
Possible related: #56 #75

**Comment by shadowspawn** [comment]
Reproduced related issue with action handler:

```
const program = require('commander');

program
  .command('sub <args...>')
  .option('-a')
  .action((args) => {
    console.log(args);
  })

program
  .parse(process.argv);
```

```
$ node badParse.js sub foo bar
[ 'foo', 'bar' ]
$ node badParse.js sub -a foo bar
[ 'foo', 'bar' ]
$ node badParse.js sub foo -a bar
[ 'bar', 'foo' ]
```

**Comment by shadowspawn** [comment]
Opened a PR to preserve argument order: #1138

**Comment by shadowspawn** [comment]
The argument order is fixed in the published pre-release of Commander v5.0.0 (#1163)

**Comment by shadowspawn** [comment]
Commander v5.0.0 has been released.

https://github.com/tj/commander.js/releases/tag/v5.0.0
